arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,600 @@
+// pixel's color occurring in the entire input image. A lower value in the
+// output image corresponds to that pixels value being less common in the
+// entire input image.
+//
+// Functions by first building a histogram of the input image, then, for each
+// pixel, looking up the value in the histogram and scaling this value by a
+// given factor.
+template < class InputType, class OutputType >
+void
+color_commonality_filter::priv
+::perform_filtering( vil_image_view< InputType > const& input,
+                     vil_image_view< OutputType >& output,
+                     color_commonality_filter_settings const& options )
+{
+  if( !std::numeric_limits< InputType >::is_integer )
+  {
+    LOG_ERROR( p->logger(), "Input must be an integer type." );
+    return;
+  }
+  if( !is_power_of_two( options.resolution_per_channel ) )
+  {
+    LOG_ERROR( p->logger(), "Input resolution must be a power of 2." );
+  }
+
+  // Set output image size
+  output.set_size( input.ni(), input.nj() );
+
+  // If we are in grid mode, simply call this function recursively
+  if( options.grid_image )
+  {
+    color_commonality_filter_settings recursive_options = options;
+    recursive_options.grid_image = false;
+
+    // Processes gridded sub-images
+    unsigned const ni = input.ni();
+    unsigned const nj = input.nj();
+
+    for( unsigned j = 0; j < options.grid_resolution_height; ++j )
+    {
+      for( unsigned i = 0; i < options.grid_resolution_width; ++i )
+      {
+        // Top left point for region
+        auto const left = ( i * ni ) / options.grid_resolution_width;
+        auto const top = ( j * nj ) / options.grid_resolution_height;
+
+        // Bottom right
+        auto const right = ( ( i + 1 ) * ni ) /
+                           options.grid_resolution_width;
+        auto const bottom = ( ( j + 1 ) * nj ) /
+                            options.grid_resolution_height;
+
+        // Formulate rect region
+        vgl_box_2d< unsigned > region{ left, top, right, bottom };
+
+        vil_image_view< InputType > region_data_ptr =
+          point_view_to_region( input, region );
+        vil_image_view< OutputType > output_data_ptr =
+          point_view_to_region( output, region );
+
+        // Process rectangular region independently of one another
+        perform_filtering( region_data_ptr, output_data_ptr,
+                           recursive_options );
+      }
+    }
+  }
+  else
+  {
+    // Reset histogram (use internal or external version)
+    unsigned int hist_size = options.resolution_per_channel;
+
+    if( input.nplanes() == 3 )
+    {
+      hist_size = options.resolution_per_channel *
+                  options.resolution_per_channel *
+                  options.resolution_per_channel;
+    }
+
+    std::vector< unsigned >* histogram = options.histogram;
+    std::unique_ptr< std::vector< unsigned > > local_histogram;
+
+    if( !histogram )
+    {
+      local_histogram.reset( new std::vector< unsigned >{ hist_size } );
```

Comment:
I think you missed:
```suggestion
      local_histogram.reset( new std::vector< unsigned >{ hist_size } );
      histogram = local_histogram.get();
```
...which is sort of important since it will SEGV otherwise :slightly_smiling_face:.

Reply:
Ack, thanks. Batch adding suggestions was breaking.